authorizations/authorization-ranger/src/main/java/org/apache/gravitino/authorization/ranger/RangerAuthorizationPlugin.java

```diff
@@ -0,0 +1,998 @@
+   * 2. If the policy exists and has the same privilege, because support idempotent operation, so
+   * return true. <br>
+   * 3. If the policy exists but has different privileges, also return true. Because one Ranger
+   * policy maybe contain multiple Gravitino securable object <br>
+   * 4. If the policy does not exist, then create a new policy. <br>
+   */
+  private boolean doAddSecurableObject(RoleChange.AddSecurableObject change) {
+    RangerPolicy policy = findManagedPolicy(change.getSecurableObject());
+
+    if (policy != null) {
+      // Check the policy item's accesses and roles equal the Gravitino securable object's privilege
+      // and roleName
+      Set<String> policyPrivileges =
+          policy.getPolicyItems().stream()
+              .filter(policyItem -> policyItem.getRoles().contains(change.getRoleName()))
+              .flatMap(policyItem -> policyItem.getAccesses().stream())
+              .map(RangerPolicy.RangerPolicyItemAccess::getType)
+              .collect(Collectors.toSet());
+      Set<String> newPrivileges =
+          change.getSecurableObject().privileges().stream()
+              .filter(Objects::nonNull)
+              .flatMap(privilege -> translatePrivilege(privilege.name()).stream())
+              .filter(Objects::nonNull)
+              .collect(Collectors.toSet());
+      if (policyPrivileges.containsAll(newPrivileges)) {
+        LOG.info(
+            "The privilege({}) already added to Ranger policy({})!",
+            policy.getName(),
+            change.getSecurableObject().fullName());
+        // If it exists policy with the same privilege, then directly return true, because support
+        // idempotent operation.
+        return true;
+      }
+    } else {
+      policy = new RangerPolicy();
+      policy.setService(rangerServiceName);
+      policy.setName(change.getSecurableObject().fullName());
+      policy.setPolicyLabels(Lists.newArrayList(MANAGED_BY_GRAVITINO));
+
+      List<String> nsMetadataObject =
+          Lists.newArrayList(
+              SecurableObjects.DOT_SPLITTER.splitToList(change.getSecurableObject().fullName()));
+      if (nsMetadataObject.size() > 4) {
+        // The max level of the securable object is `catalog.db.table.column`
+        throw new RuntimeException("The securable object than 4");
+      }
+      nsMetadataObject.remove(0); // remove `catalog`
+
+      List<String> rangerDefinesList =
+          Lists.newArrayList(
+              RangerDefines.RESOURCE_DATABASE,
+              RangerDefines.RESOURCE_TABLE,
+              RangerDefines.RESOURCE_COLUMN);
+      for (int i = 0; i < nsMetadataObject.size(); i++) {
+        RangerPolicy.RangerPolicyResource policyResource =
+            new RangerPolicy.RangerPolicyResource(nsMetadataObject.get(i));
+        policy.getResources().put(rangerDefinesList.get(i), policyResource);
+      }
+    }
+
+    addPolicyItem(policy, change.getRoleName(), change.getSecurableObject());
+    try {
```

Comment:
Maybe you can extract a method `CreateOrUpdatePolicy` in `PolicyUtil` or `PolicyHelper` class.

Reply:
This comment isn't resolved. This code is duplicate with Line 569.